### PR TITLE
Add GUI and more

### PR DESCRIPTION
**Pull request by fenhl**
This is a complete rewrite of the script in [Rust](https://rust-lang.org/). It preserves the following features of the Python script:

- [x] Command-line interface (now simply called `rsl`)
- [x] Random settings league weights
- [ ] Fully random weights (maybe as a “load preset” option in the custom tab)
- [x] Co-op weights
- [x] Multiworld weights
- [x] Custom weights
- [x] Load weights from file
- [x] Toggle to disable Standard/RSL tricks (not available for league seeds)
- [ ] Toggle to disable RSL conditionals (no longer necessary with the new custom conditionals feature)
- [x] Randomized and/or fixed starting items

It also adds some new features:

- [x] Absolutely no dependencies
    * The release script builds a standalone `.exe` for Windows and a standalone `.app` for macOS, and attaches them to a new GitHub release. There are [instructions](https://github.com/fenhl/plando-random-settings/tree/riir#releasing-a-new-version) on how to use the release script.
    * if Python is not installed, the app still runs, but shows an error message when “Generate Seed” is clicked
        * on Windows, that error message has an “Install Python” button
    * Automatically downloads/updates the randomizer
    * [Compiling from source](https://github.com/fenhl/plando-random-settings/tree/riir#building-from-source) is easy too
- [x] Graphical UI for macOS and Windows (should work on Linux too but not tested)
- [x] Command-line options (see `rsl --help`)
- [x] Enforces Dev-R version for league seeds to a specific commit
- [x] Conditionals are part of the new weights file format, allowing custom conditionals
- [x] Weights files have support for number ranges with uniform/geometric distribution, removing the hardcoded weigts for settings like `bridge_tokens`
- [x] Save custom weights to file
- [x] Self-updater

If this lands, I'm also planning to add support for different presets (and league vs current rando version) to RSLBot.

### CODE REVIEW

**Comment by fenhl** [comment]
With the exception of fully random weights, which make more sense to be added back in later, together with a more user-friendly Custom tab, this is now ready.

To publish the first release, before following the [instructions](https://github.com/fenhl/plando-random-settings/tree/riir#releasing-a-new-version), you will have to either rename the git tag of the latest release to `v6.0.12` or create a dummy release with that tag name. This is because the release script and self-updater expect release tags to be in that format.

**Comment by matthewkirby** [comment]
Are any of these files not required? Things like `cargo.lock`. If so, we should remove them

File: crate/ootr/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    quote::quote,
+    zip::ZipArchive,
+};
+
+const REPO_NAME: &str = "OoT-Randomizer";
+const LEAGUE_COMMIT_HASH: &str = "b670183e9aff520c20ac2ee65aa55e3740c5f4b4";
+const LEAGUE_VERSION: &str = "5.2.117 R-1";
+
+fn import<'p>(py: Python<'p>, module: &str) -> PyResult<&'p PyModule> {
+    let project_dirs = ProjectDirs::from("net", "Fenhl", "RSL").expect("missing home directory");
```
Comment (matthewkirby) [inline crate/ootr/src/lib.rs:30]: is this a hardcoded path?

File: crate/rsl-gui/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+version = "0.99"
+default-features = false
+features = ["from"]
+
+[dependencies.enum-iterator]
+git = "https://github.com/dasgefolge/enum-iterator"
+branch = "fields" #TODO finish this PR
+
+[dependencies.iced]
+git = "https://github.com/hecrj/iced" #TODO go back to the version on crates.io once it's compatible with tokio 1
```
Comment (matthewkirby) [inline crate/rsl-gui/Cargo.toml:31]: Two TODO's here. Are these still relevant?

**Review by matthewkirby — CHANGES_REQUESTED** [review]
- [x] How many of these things are generated files that do not need to be included in the repo?
- [x] In the README under releasing new versions, would be good to also include instructions to bump the rando version. Is it just updating `LEAGUE_COMMIT_HASH` and `LEAGUE_VERSION` in `crate/ootr/src/lib.rs`? Also instructions for adding new conditional settings.
- [ ] It seems as though there is a lot of repeated stuff in the various `Cargo.toml` files. Why do we need so many of them?
- [ ] Quite a few `TODO`s in `crate/rsl-gui/src/main.rs` Might be good to close those out, take them out, or open up an issue of future work.

**Comment by matthewkirby** [comment]
I also tested a build from source and got this error
image

**Comment by fenhl** [comment]
> Are any of these files not required? Things like `cargo.lock`. If so, we should remove them

Cargo.lock ensures reproducible builds by locking the exact versions of dependencies. The Rust team [recommends](https://doc.rust-lang.org/stable/cargo/guide/cargo-toml-vs-cargo-lock.html) not gitignoring it for binaries.

> * [ ]  How many of these things are generated files that do not need to be included in the repo?

All generated files other than `Cargo.lock` are in the `target` directory, which is ignored.

> * [ ]  In the README under releasing new versions, would be good to also include instructions to bump the rando version. Is it just updating `LEAGUE_COMMIT_HASH` and `LEAGUE_VERSION` in `crate/ootr/src/lib.rs`? Also instructions for adding new conditional settings.

In most cases, that should be all, yes. Will add docs, kind of forgot about that.

> * [ ]  It seems as though there is a lot of repeated stuff in the various `Cargo.toml` files. Why do we need so many of them?

This is a consequence of me structuring the project as several separate crates (Rust packages). Each crate has its own Cargo.toml listing its dependencies. The crates are:

* `ootr`: This pulls some data from the Python code, like the location and trick lists. This is a macro crate, which means it's run at compile time, not at runtime. This way, the app itself does not require Python to launch (except when the “Generate Seed” button is clicked of course).
* `rsl`: Code shared by the CLI, GUI, and release script.
* `rsl-cli`: The command-line frontend.
* `rsl-gui`: The main app.
* `rsl-updater`: The self-updater for Windows (required because Windows does not allow deleting running programs).
* `rsl-utils`: Contains the release script.

Cargo does allow having multiple binaries in the same crate, but since `ootr` has to be a separate crate, I opted to organize the binaries at the crate level to avoid having two layers of organization in the project.

> * [ ]  Quite a few `TODO`s in `crate/rsl-gui/src/main.rs` Might be good to close those out, take them out, or open up an issue of future work.

A lot of them should be resolved by the “more user-friendly Custom tab” feature, for which I'll open a separate issue. I'll make sure to go through the rest and fix/open an issue for them as well.

> I also tested a build from source and got this error
> image

I forgot about this. `cargo install --git` doesn't support multi-crate workspaces, so building from source needs to be done as:

```sh
git clone https://github.com/fenhl/plando-random-settings --branch=riir
cargo install --path=plando-random-settings/crate/rsl-cli
```

(installing the CLI from my branch as an example)

File: crate/ootr/src/lib.rs (hunk 1 above)
Comment (fenhl) [reply crate/ootr/src/lib.rs:30]: Sort of. This uses the [directories](https://docs.rs/directories) crate to save things at the idiomatic location for each platform, e.g. `~\AppData\Local\Fenhl\RSL\cache` on Windows. This project uses that directory for the copies of the randomizer (latest/league), user settings (config.json, currently only whether to check for updates on launch), the plando and randomizer settings file that are saved here before running the randomizer, and the Python installer if the uses clicks the “Python not found — Install” button.

File: crate/rsl-gui/Cargo.toml (hunk 2 above)
Comment (fenhl) [reply crate/rsl-gui/Cargo.toml:31]: The first one is a reference to <https://github.com/stephaneyfx/enum-iterator/pull/7> which @leiftw and I still need to finish. The second is currently waiting on a new [iced](https://docs.rs/iced) release — that crate has historically been slow to push changes from GitHub to crates.io, but the maintainer [plans to release more frequently](https://github.com/hecrj/iced/issues/706#issuecomment-763718669).

**Comment by matthewkirby** [comment]
New errors, yay! 
image

**Comment by fenhl** [comment]
Python should normally be in PATH as `python`. Since you don't seem to have that and are using the launcher, you can try running `set PYO3_PYTHON=py` before running the `cargo` command. If that doesn't work, you may have to install Python 3.9 from the official website, making sure to check the “add to PATH” option.

**Comment by matthewkirby** [comment]
Hmm weird that my path wasn't set up correctly. When I installed python I just used the windows installer. But yeah that worked

**Comment by matthewkirby** [comment]
Next up, lol
image

**Comment by fenhl** [comment]
Uhh what? That one doesn't make any sense. Can you check `rustc --version` and make sure you're on 1.49.0?

**Comment by matthewkirby** [comment]
Yeah, version is correct. I ran `cargo clean` in the repo directory and ran the install again to make sure something didn't get messed up while it was erroring before. Restarted the compile and got the same error

**Comment by matthewkirby** [comment]
Does the code still need to be with the rando code when installing from source?

**Comment by fenhl** [comment]
No, it will download the league copy to the cache directory. Are you on Windows 10? The only thing I can think of is maybe older versions of Windows don't work well with `/` rather than `\` in paths. If that's not the issue, maybe I can send you an invite for the Rust community Discord server to debug this?

**Comment by matthewkirby** [comment]
I am up to date and on windows 10. Honestly, I don't really understand the code so debugging it in the rust server probably wouldnt be too useful.

**Comment by fenhl** [comment]
So this might be another Python issue. Can you try and check if it goes away by installing Python from the website (with PATH addition) and resetting `PYO3_PYTHON` to `python`? Otherwise, can you check if a similar error (os error 126) happens if you run `cargo build --package=ootr`?

**Comment by matthewkirby** [comment]
Alrighty. So here's what happened
- The default python installer doesnt add python to path. When I installed python awhile back I probably missed the checkbox.
- Because of that, I was using cmd and not power shell since in cmd `py --version` gave me an output
- cmd is even derpier than powershell. 
- I reinstalled python and selected the add to path option this time then did the cargo install in powershell and it worked.

**Comment by matthewkirby** [comment]
You used `rsl-cli` as an example above. Will that install everything or do I still need to `cargo install` the other crates?

**Comment by fenhl** [comment]
You can also replace `rsl-cli` with `rsl-gui` to build the GUI from source, then run it as `rsl-gui`. Those are the only two useful crates.

I've updated the readme, which should address your points except for the TODOs.

**Comment by matthewkirby** [comment]
Yeah I went through and ran them each and am getting another error in `rsl-gui`
image

**Comment by fenhl** [comment]
Right, that's the self-updater `.exe` being embedded into the GUI `.exe`. That should probably be optional.

**Comment by matthewkirby** [comment]
That directory just doesn't exist. 
image

**Comment by fenhl** [comment]
You can create it using `cargo build --release --target=x86_64-pc-windows-msvc --package=rsl-updater`. The release script runs this command before compiling `rsl-gui`, but it shouldn't be necessary for compiling from source, since if you're doing that you're probably not going to use the self-updater.

**Comment by matthewkirby** [comment]
Alright. We should include that step in the install from source instructions though. Otherwise we have instructions that don't work. I need to head out but I will be back in a couple hours to finish up installing and testing.

**Comment by fenhl** [comment]
I'm just going to make the self-updater an optional feature that's enabled by the release script, so that step isn't required.

**Comment by matthewkirby** [comment]
Alrighty, just did a fresh install from source on windows and it looks great!. I have some additional comments though

## The custom tab
- [ ] The custom tab is very hard to understand what is going on. This is just a formatting thing though. Putting more pronounced headings would go a great distance I think (for things like separating disabled locations and allowed tricks).
- [ ] For each setting, it looks like I can change the name and setting name. but in reality, this will actually break and cause the rom to not generate. These could just be fixed and only changed by adding new settings or new settings value. 
- [ ] On that same front, for settings names, option names and weights, the text and numbers are too light I think. I thought they were suggestion text (like the text that might say "enter response here" or something).

## Other comments
- [ ] Maybe put a border around the content that shows up when you select one of the radio buttons at the top? It is hard to follow what is going on when you scroll since there is no indication of where the scroll box is
- [ ] Very small thing, but is it possible to make the boxes for the paths and typing a bit bigger? Just add a bit of padding from the text and border. We also probably want it vertically centered with the browse or other buttons adjacent to them

**Comment by fenhl** [comment]
> * [ ]  The custom tab is very hard to understand what is going on. This is just a formatting thing though. Putting more pronounced headings would go a great distance I think (for things like separating disabled locations and allowed tricks).

I made some changes here. Rather than setting the headings in bold (which would require loading a bold font from the system — something I haven't done before) I put rulers between each option. Do you think that works?

> * [ ]  For each setting, it looks like I can change the name and setting name. but in reality, this will actually break and cause the rom to not generate. These could just be fixed and only changed by adding new settings or new settings value.

Yeah, that's a big part of the “more user-friendly Custom tab” feature, which I think is probably going to take long enough that it makes sense to release this as-is for now and add the feature in a separate update. I've opened #11 for this now.

> * [ ]  On that same front, for settings names, option names and weights, the text and numbers are too light I think. I thought they were suggestion text (like the text that might say "enter response here" or something).

Changed to black text. Not sure why that's not the default.

> * [ ]  Maybe put a border around the content that shows up when you select one of the radio buttons at the top? It is hard to follow what is going on when you scroll since there is no indication of where the scroll box is

Done.

> * [ ]  Very small thing, but is it possible to make the boxes for the paths and typing a bit bigger? Just add a bit of padding from the text and border. We also probably want it vertically centered with the browse or other buttons adjacent to them

Done.

**Comment by matthewkirby** [comment]
> > The custom tab is very hard to understand what is going on. This is just a formatting thing though. Putting more pronounced headings would go a great distance I think (for things like separating disabled locations and allowed tricks).

> I made some changes here. Rather than setting the headings in bold (which would require loading a bold font from the system — something I haven't done before) I put rulers between each option. Do you think that works?

I think it looks fine for now. This is something we can work on in issue #11 though. 

I agree that having the custom tab in #11 is probably good enough for now and not necessary for release. All of the other stuff looks wayyy better imo. 

From here, I think I just want to test building on macos and preparing a release. Then its probably good to go

**Comment by fenhl** [comment]
Do you think it makes sense to change the version number to 7.0.0 or maybe 2.0.0? Since we won't always be able to have it match the current Dev-R version: When releasing an update but Dev-R hasn't changed (or only the R-n part), we'll have to increment the version number anyway to make the self-updater detect it.

**Comment by matthewkirby** [comment]
I think 2.0.0 is probably fine. I wasn't using semantic versioning before because I was just tracking the version of the randomizer. Since we don't need to do that anymore, going to v2 is probably better than 7

**Comment by fenhl** [comment]
The racetime.gg chat bot is now part of this PR too. The new version has support for `solo` and `coop` presets in addition to the default `league`. I've done some basic tests and everything seems to be working.

**Comment by matthewkirby** [comment]
I’m not sure that is appropriate for this repo? That means everyone who
wants to get the GUI now has the bot as well?

**Comment by fenhl** [comment]
That's not how it works. Even if you build using `cargo install`, it only installs the part you actually specify. I think having the bot in here makes sense since it's pretty much just a wrapper around the `rsl::generate` function.

File: README.md [hunk 3]
```diff
@@ -1,21 +1,111 @@
...
+    ```sh
+    cargo install --path=crate/rsl-gui
+    ```
+7. You can now launch the app by running the command `rsl-gui`, and/or use the command-line version `rsl` (see `rsl --help` for details).
+
+## Releasing a new version
+
+You will need a 64-bit Windows PC and a Mac.
+
+1. If the randomizer version for league seeds should be updated, change the constants `LEAGUE_COMMIT_HASH` and `LEAGUE_VERSION` in `crate/ootr/src/lib.rs` accordingly
```
Comment (SlyryD) [inline README.md:79]: Can you elaborate on what the LEAGUE_COMMIT_HASH is? I assume that's the commit that you push to the repo?

File: README.md [hunk 4]
```diff
@@ -1,21 +1,111 @@
...
+2. Bump the version number in:
+    * `assets/macos/RSL.app/Contents/Info.plist` (also increment `CFBundleVersion`)
```
Comment (SlyryD) [inline README.md:81]: Can you call out both CFBundleShortVersionString and CFBundleVersion? Do you update CFBundleVersion by adding 1?

File: README.md (hunk 3 above)
Comment (fenhl) [reply README.md]: It's actually the Dev-R commit.

File: README.md [hunk 5]
```diff
@@ -1,21 +1,112 @@
...
+1. If the randomizer version for league seeds should be updated, change the constants `LEAGUE_COMMIT_HASH` and `LEAGUE_VERSION` in `crate/ootr/src/lib.rs` accordingly. `LEAGUE_COMMIT_HASH` is the commit hash for the Dev-R version to be used, which can be copied using the clipboard button at <https://github.com/Roman971/OoT-Randomizer/commits/Dev-R>, and `LEAGUE_VERSION` is the contents of the `__version__` variable in `version.py` in that commit.
+2. Bump the version number in:
+    * `assets/macos/RSL.app/Contents/Info.plist` (change `CFBundleShortVersionString` to the new version number and increment `CFBundleVersion` by 1)
+    * `crate/ootr/Cargo.toml`
+    * `crate/rsl/Cargo.toml`
+    * `crate/rsl-cli/Cargo.toml`
+    * `crate/rsl-gui/Cargo.toml`
+    * `crate/rsl-updater/Cargo.toml`
+    * `crate/rsl-utils/Cargo.toml`
+    * `crate/rslbot/Cargo.toml`
```
Comment (SlyryD) [inline README.md:88]: Is it just me, or is this not showing up in the markdown preview?

File: README.md [hunk 6]
```diff
@@ -1,21 +1,112 @@
...
+        * `"macRepoPath"`: The path to the clone of this repo on your Mac (optional, if not given no macOS app is included in the release)
+        * `"releaseNotesEditor"`: A command that will be called with `--wait` and the path to an empty file for writing the release notes (optional, defaults to `"C:\\Program Files\\Microsoft VS Code\\bin\\code.cmd"`)
+        * `"repoOwner"`: The GitHub user or org where the release should be published (optional, defaults to `"matthewkirby"`)
+        * `"repoName"`: The name of the repository where the release should be published (optional, defaults to `"plando-random-settings"`)
+    4. In the repo, run `cargo run --release --package=rsl-utils --bin=rsl-release`. You can add the following command-line options, but must separate them with a `--` (so they're interpreted as `rsl-release` options, not `cargo` options):
+        * To see output of build commands as they are happening, use `--verbose`. Note that this may considerably slow down the release script since it won't run multiple build commands in parallel.
+        * To create the GitHub release as a draft but not publish it, use `--no-publish`.
```
Comment (SlyryD) [inline README.md:112]: I couldn't get the --no-publish flag to work. Can you show an example? Is it this?

cargo run --release --package=rsl-utils --bin=rsl-release -- --verbose -- --no-publish

File: README.md (hunk 5 above)
Comment (fenhl) [reply README.md:88]: It shows up for me. I just added it in the latest commit, maybe you're looking at the previous one?

File: README.md (hunk 6 above)
Comment (fenhl) [reply README.md:112]: The `--` between `--verbose` and `--no-publish` is the issue here. I should clarify that.

File: README.md [hunk 7]
```diff
@@ -1,21 +1,112 @@
...
+        * `"githubToken"`: The personal access token from step 1
+        * `"macHostname"`: A hostname or IP address (and if necessary the username, in `"username@hostname"` form) at which your Mac is reachable via SSH (optional, if not given no macOS app is included in the release)
+        * `"macRepoPath"`: The path to the clone of this repo on your Mac (optional, if not given no macOS app is included in the release)
+        * `"releaseNotesEditor"`: A command that will be called with `--wait` and the path to an empty file for writing the release notes (optional, defaults to `"C:\\Program Files\\Microsoft VS Code\\bin\\code.cmd"`)
+        * `"repoOwner"`: The GitHub user or org where the release should be published (optional, defaults to `"matthewkirby"`)
+        * `"repoName"`: The name of the repository where the release should be published (optional, defaults to `"plando-random-settings"`)
+    4. In the repo, run `cargo run --release --package=rsl-utils --bin=rsl-release`. You can add the following command-line options, but must separate them with a `--` (so they're interpreted as `rsl-release` options, not `cargo` options):
```
Comment (SlyryD) [inline README.md]: I haven't been able to get a successful release. It's a rustc error apparently because the x86_64-pc-windows-mvc release target is missing.

Here's my [branch](https://github.com/SlyryD/plando-random-settings).

Here's my [release-config.json](https://github.com/matthewkirby/plando-random-settings/files/6041462/release-config.json.txt) (without token)

Here's the [output](https://github.com/matthewkirby/plando-random-settings/files/6041445/cargolog.txt) of the command.

Comment (fenhl) [reply README.md]: That was caused by a typo in the release script. Apparently I forgot to test this after adding the updater. Try again please.

Comment (fenhl) [reply README.md]: Tested a release, seems to be working again.

**Comment by TreZc0** [comment]
I don't want to play the party pooper here...but why the rewrite in rust?
From a meta standpoint, it makes no sense.
It moves to different langugage with a tiny percentage of knowledge base in the entire community, it puts the entire code maintaining responsibility on basically one person, and it pushes it even further away from the randomizer.

I think this overall is a bad decision.

**Comment by fenhl** [comment]
Addition to the main rando was rejected by the dev team. The rewrite is to make it more accessible for users by building a standalone executable. Accessibility for users is more important than accessibility for devs.

**Comment by TreZc0** [comment]
I don't want to discredit your work or anything you have done, but sorry, I disagree with that reason.
A full rust rewrite, while complicating things for dev work apart from you, is not at all a requirement for a standalone executable.
As for the rejection by the dev team, I am sure we can work something out, but that's of course not a requirement, just a suggestion.
All of this are just my thoughts, but I imagine that others share my concerns.

**Comment by fenhl** [comment]
Sure, you could use C or C++ instead, but I really don't think that would be better.

**Comment by TreZc0** [comment]
Or stick with python and use pyexecutable, which is what i would suggest.

**Comment by fenhl** [comment]
[This](https://pypi.org/project/pyexecutable/)? Is there any further documentation?

**Comment by matthewkirby** [comment]
For what its worth, my original plan was to write the gui in react and build an electron app. But I couldn't figure out how to get electron working haha. At the end of the day, python UIs still look kinda gross. I also wrote the GUI already but put it down when fenhl offered this.

**Comment by leiftw** [comment]
Is there something inherent about Python that makes them look gross?

**Comment by matthewkirby** [comment]
So I actually looked into it again after posting that and I take that back.
I remember (incorrectly it seems) that python didn’t really have any great
ways to format GUIs. I remember really only seeing super blocky buttons
that you couldn’t format at all.

Now that I’ve looked into it though I think we could (and I would happily
lead the effort) make a good looking, functional GUI while sticking to
python. I certainly would be able to do much much more than I could in rust
(although that’s not really saying much I guess).

**Comment by matthewkirby** [comment]
I just wanted to check in and make note that I have begun working on an implementation entirely in python.

**Comment by fenhl** [comment]
Do you plan to add the new features from this version? In particular:

* Dev-R version management is required for me to continue to run RSLBot
* A way to specify different weights without editing the script (ideally via Python API but a CLI would work too) is required for me to add support for different presets to RSLBot
* A way to specify world count for multiworld weights is required for me to add support for multiworld weights to RSLBot

**Comment by matthewkirby** [comment]
> Dev-R version management is required for me to continue to run RSLBot

Yes 100%

> A way to specify different weights without editing the script (ideally via Python API but a CLI would work too) is required for me to add support for different presets to RSLBot

I can do this via a CLI pretty easily. 

> A way to specify world count for  multiworld weights is required for me to add support for multiworld weights to RSLBot

This can also be through the CLI
